### ISSUE TITLE
Incorrect handling of concurrent paremetrized UPDATE queries

### ISSUE THREAD

**Issue by kheaactua** [google/lovefield#136]
When I create an update query with bound values, there `where` portion doesn't work.  But when I recreate the query every time, it's fine.

i.e. The following code
```javascript
var eventAvgQuery = db.update(event_table)
	.set(event_table.avgHeight, lf.bind(0))
	.set(event_table.avgScore,  lf.bind(1))
	.where(event_table.id.eq(lf.bind(2)));

var vals = [0.1, 0.2, 2];
eventAvgQuery.bind(vals).exec();
```
causes every row in the event table to be updated, regardless of it's id,

while
```javascript
var vals = [0.1, 0.2, 2];
db.update(event_table)
	.set(event_table.avgHeight, vals[0])
	.set(event_table.avgScore,  vals[1])
	.where(event_table.id.eq(vals[2])).exec();
```
executes properly.

I've double checked the values being bounded, made sure it wasn't a concurrency thing, etc.

**Comment by freshp86** [comment]
I think you are misusing the parametrized query API, see https://github.com/google/lovefield/blob/master/docs/spec/04_query.md#45-parameterized-query.

In your code you need to call `bind()` before calling `exec()`. Also passing the array with the bound values to `exec()` seems wrong.

**Comment by kheaactua** [comment]
Ack, that was a typo for when I was writing it into the issue, I'll edit it to be proper.  (also went through my undo tree to ensure I was calling bind before exec.)

**Comment by freshp86** [comment]
I tried to reproduce this without success, you can see (and tr)y my example at https://github.com/google/lovefield/commit/a4e537476fa9cd14a627bc91c2b6f7f795cbfec1. Also could you post the query execution plan in your case? You can do this by calling explain() on your query, after bind() and before exec().

**Comment by kheaactua** [comment]
Certainly,

When I create a new query per update, i.e.

```javascript
var updates = [[0.1, 0.2, 2], [0.2, 0.3, 3], [0.3, 0.4, 4];]

updates.forEach(function(vals) {
	var query = db.update(event_table)
		.set(event_table.avgHeight, vals[0])
		.set(event_table.avgScore,  vals[1])
		.where(
			event_table.id.eq(vals[2])
		);
	promises.push(query.exec().then(function() { console.debug(query.explain());}));
});
return Promise.all(promises).then();
```

I get explanations like (note, the data in `updates` above isn't what's being used in the actual queries)
```
update(Event)
-table_access_by_row_id(Event)
--index_range_scan(Event.pkEvent, [1, 1], natural)

update(Event)
-table_access_by_row_id(Event)
--index_range_scan(Event.pkEvent, [2, 2], natural)

update(Event)
-table_access_by_row_id(Event)
--index_range_scan(Event.pkEvent, [3, 3], natural)
```

As opposed to having one query, i.e
```javascript

// Real data is used in this case
var query = db.update(event_table)
	.set(event_table.avgHeight, lf.bind(0))
	.set(event_table.avgScore,  lf.bind(1))
	.where(
		event_table.id.eq(lf.bind(3)) // With the real data, I use the third element rather than the second as the key
	);

updates.forEach(function(vals) {
	console.log(vals);
	console.log(query.explain(query.bind(vals)));

	promises.push(query.bind(vals).exec().then(function() { console.debug(query.explain());}));
});
return Promise.all(promises).then();
```

I get the explanations
```
[0.46682994375005366, 0.6222222222222222, 1, 1]
update(Event)
-table_access_by_row_id(Event)
--index_range_scan(Event.pkEvent, [1, 1], natural)

[0.869320666088414, 0.7555555555555555, 1, 2]
update(Event)
-table_access_by_row_id(Event)
--index_range_scan(Event.pkEvent, [2, 2], natural)

[1.4190526150830896, 0.7555555555555555, 1, 3]
update(Event)
-table_access_by_row_id(Event)
--index_range_scan(Event.pkEvent, [2, 2], natural)

// ...

update(Event)
-table_access_by_row_id(Event)
--index_range_scan(Event.pkEvent, [10, 10], natural)

// This one is repeated 10 times
update(Event)
-table_access_by_row_id(Event)
--index_range_scan(Event.pkEvent, [10, 10], natural)
```

In the first case, when looking at the data through the inspector, all my `avgHeight`'s and `avgScore`'s were unique, but in the second case, they're all set to the same value.

I did one more test, just to eliminate the possibility of this being a concurrency issue...  while I thought I tested for this yesterday, I think it might be that:

```javascript
var query = db.update(event_table)
	.set(event_table.avgHeight, lf.bind(0))
	.set(event_table.avgScore, lf.bind(1))
	.where(
		event_table.id.eq(lf.bind(3))
	);

console.log(updates[0]);
console.log(query.bind(updates[0]));
query.bind(updates[0]).exec().then(function() { console.debug(query.explain());})
	.then(function() {
		console.log(updates[1]);
		console.log(query.bind(updates[1]));
		return query.bind(updates[1]).exec().then(function() { console.debug(query.explain());});
	})
	.then(function() {
		console.log(updates[2]);
		console.log(query.bind(updates[2]));
		return query.bind(updates[2]).exec().then(function() { console.debug(query.explain());})
	});
```

yields,
```
Object {status: 1, name: "E. Washington", lat: 38.48414701233152, lon: -86.03122873392189, volAdded: 2605.2048711571842…}
[0.4917334748742481, 0.8666666666666667, 1, 1]
lf.q…y.UpdateBuilder {global: lf.Global, queryEngine_: lf.p…c.DefaultQueryEngine, runner_: lf.p…c.Runner, query: lf.q…y.UpdateContext}
Site E. Washington populated
All sites created
update(Event)
-table_access_by_row_id(Event)
--index_range_scan(Event.pkEvent, [1, 1], natural)

[0.8107417736171452, 0.7555555555555555, 1, 2]
lf.q…y.UpdateBuilder {global: lf.Global, queryEngine_: lf.p…c.DefaultQueryEngine, runner_: lf.p…c.Runner, query: lf.q…y.UpdateContext, plan_: lf.p…c.PhysicalQueryPlan}
update(Event)
-table_access_by_row_id(Event)
--index_range_scan(Event.pkEvent, [2, 2], natural)

[1.4804333344610592, 0.9111111111111111, 1, 3]
lf.q…y.UpdateBuilder {global: lf.Global, queryEngine_: lf.p…c.DefaultQueryEngine, runner_: lf.p…c.Runner, query: lf.q…y.UpdateContext, plan_: lf.p…c.PhysicalQueryPlan}
update(Event)
-table_access_by_row_id(Event)
--index_range_scan(Event.pkEvent, [3, 3], natural)
```

and checking the DB, the values are set properly.  Looks like I might be binding wrong with my promises?

**Comment by freshp86** [comment]
Thanks for providing more info. I have a theory about what might be happening

calling bind() before the previous bound query has finished executing seems to be messing up the query that is in flight. Can you try the following?

```
var updates = [[0.1, 0.2, 2], [0.2, 0.3, 3], [0.3, 0.4, 4];
var query = db.update(event_table)
        .set(event_table.avgHeight, lf.bind(0))
        .set(event_table.avgScore,  lf.bind(1))
        .where(event_table.id.eq(lf.bind(2));

var runQuery = function(index) {
  return q.bind(updates[index]).exec();
};
runQuery(0).then(runQuery[1]).then(runQuery[2]);
```

Does that produce the correct result?

**Comment by freshp86** [comment]
Thanks for uncovering this very subtle bug. We have a fix on the way. Will update this thread once the fix has landed.

**Comment by arthurhsu** [comment]
Fixed 00b9acf85de30a869d2386ebc25b825c63eef8df

**Comment by kheaactua** [comment]
Hi,

Sorry for the late reply, your code suggestion worked.  (I haven't tried using the updated lib yet though)

Specifically:
```javascript

// Real data is used in this case
var query = db.update(event_table)
	.set(event_table.avgHeight, lf.bind(0))
	.set(event_table.avgScore,  lf.bind(1))
	.where(
		event_table.id.eq(lf.bind(3)) // With the real data, I use the third element rather than the second as the key
	);

var runQuery = function(index) {
	var a = query.bind(updates[index]);
	console.log(query.explain());
	return query.bind(updates[index]).exec().then(function() { console.log(query.explain()); });
}

updates.forEach(function(vals.idx) {
	promises.push(runQuery(idx);
});
return Promise.all(promises).then();
```

Yielded: 
```
update(Event)
-table_access_by_row_id(Event)
--index_range_scan(Event.pkEvent, [1, 1], natural)

9x update(Event)
-table_access_by_row_id(Event)
--index_range_scan(Event.pkEvent, [2, 2], natural)

10x update(Event)
-table_access_by_row_id(Event)
--index_range_scan(Event.pkEvent, [10, 10], natural)
```

The keys in the explanation are weird, but when I check the database, the values are written properly.

I'll try to update my version of lovefield and use a prepared query.  Thanks for the help!
